pkg/process/ptrace/ptrace_linux_amd64.go

```diff
@@ -0,0 +1,115 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ptrace
+
+import (
+	"encoding/binary"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+const syscallInstrSize = 2
+
+func getIp(regs *syscall.PtraceRegs) uintptr {
+	return uintptr(regs.Rip)
+}
+
+func getRegs(pid int, regsout *syscall.PtraceRegs) error {
+	err := syscall.PtraceGetRegs(pid, regsout)
+	if err != nil {
+		return errors.Wrapf(err, "get registers of process %d", pid)
+	}
+
+	return nil
+}
+
+func setRegs(pid int, regs *syscall.PtraceRegs) error {
+	err := syscall.PtraceSetRegs(pid, regs)
+	if err != nil {
+		return errors.Wrapf(err, "set registers of process %d", pid)
+	}
+
+	return nil
+}
+
+// Syscall runs a syscall at main thread of process
+func (p *TracedProgram) Syscall(number uint64, args ...uint64) (uint64, error) {
+	// save the original registers and the current instructions
+	err := p.Protect()
+	if err != nil {
+		return 0, err
+	}
+
+	var regs syscall.PtraceRegs
+
+	err = getRegs(p.pid, &regs)
+	if err != nil {
+		return 0, err
+	}
+	// set the registers according to the syscall convention. Learn more about
+	// it in `man 2 syscall`. In x86_64 the syscall nr is stored in rax
+	// register, and the arguments are stored in rdi, rsi, rdx, r10, r8, r9 in
+	// order
+	regs.Rax = number
+	for index, arg := range args {
+		// All these registers are hard coded for x86 platform
+		if index == 0 {
+			regs.Rdi = arg
+		} else if index == 1 {
+			regs.Rsi = arg
+		} else if index == 2 {
+			regs.Rdx = arg
+		} else if index == 3 {
+			regs.R10 = arg
+		} else if index == 4 {
+			regs.R8 = arg
+		} else if index == 5 {
+			regs.R9 = arg
+		} else {
+			return 0, errors.New("too many arguments for a syscall")
+		}
```

Comment:
instad of looping over the params, could they be named? Or maybe better, put into their own struct?

Reply:
I agree that those if statements can be improved. I am not sure there is a lot that can be shared across CPU architectures. `syscall.PtraceRegs` looks very different on amd64 and arm64.
The convertion from a list of arguments to specific CPU registers must be done in the `ptrce_linux_<cpu>.go` files otherwise we will have to add additional code to check the current architecture and apply different logic accordingly (which we get today by leveraging Go's linker).

What do you think about creating a shared interface like:
```
type CPURegsSetter interface {
	SetFirstArg(uint64)
	SetSecondArg(uint64)
	SetThirdArg(uint64)
	GetRegs() *syscall.PtraceRegs
}
``` 
and then implementing it differently for every architecture.

Also, I think because the [order of the registers](https://en.wikipedia.org/wiki/X86_calling_conventions#x86-64_calling_conventions) is something very hard-coded it may be enough to simply replace the if-statements with switch-case.

Reply:
The interface is a cool idea, it would be interesting to play around with this a bit more to see if we can refactor it with something like that.

But, I don't need to block on this PR over that. Just something to think about for future releases